File: Data Structures and Algorithms/AI/Regression and Gradient Descent/RMSProp/Cpp/rmsprop.cpp
#include <vector>
#include <cmath>
#include <stdexcept>
#include <type_traits>

/**
 * @brief Performs a single RMSProp optimization step.
 * 
 * @tparam T Floating point type (float, double, etc.)
 * @param params         Parameters to be updated (modified in-place).
 * @param gradients      Gradients for each parameter.
 * @param cache          Running average of squared gradients (modified in-place).
 * @param learning_rate  Step size (default: 0.001).
 * @param rho            Decay rate for the cache (default: 0.9).
 * @param epsilon        Small constant to prevent division by zero (default: 1e-8).
 * 
 * @throws std::invalid_argument If sizes of params and gradients don't match.
 */
template<typename T>
typename std::enable_if<std::is_floating_point<T>::value, void>::type
rmsprop(std::vector<T>& params,
        const std::vector<T>& gradients,
        std::vector<T>& cache,
        T learning_rate = static_cast<T>(0.001),
        T rho = static_cast<T>(0.9),
        T epsilon = static_cast<T>(1e-8))
{
    if (params.size() != gradients.size()) {
        throw std::invalid_argument("params and gradients must be the same size");
    }

    if (cache.empty()) {
        cache.resize(params.size(), T(0));
    } 
    else if (cache.size() != params.size()) {
        throw std::invalid_argument("cache must be empty or same size as params");
    }

    for (size_t i = 0; i < params.size(); ++i) {
        cache[i] = rho * cache[i] + (static_cast<T>(1) - rho) * gradients[i] * gradients[i];
        params[i] -= learning_rate * gradients[i] / (std::sqrt(cache[i]) + epsilon);
    }
}

File: Data Structures and Algorithms/AI/Regression and Gradient Descent/RMSProp/Python/rmsprop.py
import numpy as np

def rmsprop(params, gradients, cache, learning_rate=0.001, decay_rate=0.9, epsilon=1e-8):
    """    
    @param params        np.ndarray: Parameters to be updated.
    @param gradients     np.ndarray: Gradients of the parameters.
    @param cache         np.ndarray: Cache for moving average of squared gradients.
    @param learning_rate float: Learning rate (default: 0.001).
    @param decay_rate    float: Decay rate for moving average (default: 0.9).
    @param epsilon       float: Small value to prevent division by zero (default: 1e-8).

    @return tuple: A tuple containing:
        - Updated parameters (np.ndarray)
        - Updated cache (np.ndarray)
    """
    # Initialize cache if empty
    if cache.size == 0:
        cache = np.zeros_like(params)
    
    # Update cache: cache = decay * cache + (1 - decay) * gradient^2
    cache = decay_rate * cache + (1 - decay_rate) * np.square(gradients)
    
    # Update parameters: param -= learning_rate * gradient / (sqrt(cache) + epsilon)
    params = params - learning_rate * gradients / (np.sqrt(cache) + epsilon)
    
    return params, cache
